src/libraries/Common/src/Interop/Unix/System.Native/Interop.GNameUName.cs

```diff
@@ -0,0 +1,43 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Runtime.InteropServices;
+using System.Buffers;
+using System.Text;
+using System;
+using System.Collections.Generic;
+using System.Reflection;
+
+internal static partial class Interop
+{
+    internal static partial class Sys
+    {
+        /// <summary>
+        /// Gets the group name associated to the specified group ID.
+        /// </summary>
+        /// <param name="gid">The group ID.</param>
+        /// <returns>On success, return a string with the group name. On failure, throws an IOException.</returns>
+        internal static string GetGName(uint gid)
+        {
+            string result = GetGNameInternal(gid);
+            return (result == null) ? throw GetIOException(GetLastErrorInfo()) : result;
+        }
+
+        /// <summary>
+        /// Gets the user name associated to the specified user ID.
+        /// </summary>
+        /// <param name="uid">The user ID.</param>
+        /// <returns>On success, return a string with the user name. On failure, throws an IOException.</returns>
+        internal static string GetUName(uint uid)
+        {
+            string result = GetUNameInternal(uid);
+            return (result == null) ? throw GetIOException(GetLastErrorInfo()) : result;
+        }
+
+        [LibraryImport(Libraries.SystemNative, EntryPoint = "SystemNative_GetGName", StringMarshalling = StringMarshalling.Utf8, SetLastError = true)]
+        private static unsafe partial string GetGNameInternal(uint uid);
```

Comment:
Does the StringMarshalling know that it should free the `char*` that comes back from the native method?

Reply:
From looking at the generated code in ILSpy, it looks like it does:

```C#
[LibraryImport(/*Could not decode attribute arguments.*/)]
[GeneratedCode("Microsoft.Interop.LibraryImportGenerator", "42.42.42.42")]
private unsafe static string GetGroupNameInternal(uint uid)
{
	//IL_000b: Unknown result type (might be due to invalid IL or missing references)
	byte* __retVal_gen_native = default(byte*);
	Utf8StringMarshaller __retVal_gen_native__marshaller = default(Utf8StringMarshaller);
	int __lastError;
	string __retVal;
	try
	{
		Marshal.SetLastSystemError(0);
		__retVal_gen_native = __PInvoke__(uid);
		__lastError = Marshal.GetLastSystemError();
		((Utf8StringMarshaller)(ref __retVal_gen_native__marshaller)).FromNativeValue(__retVal_gen_native);
		__retVal = ((Utf8StringMarshaller)(ref __retVal_gen_native__marshaller)).ToManaged();
	}
	finally
	{
		((Utf8StringMarshaller)(ref __retVal_gen_native__marshaller)).FreeNative();  // <-- HERE
	}
	Marshal.SetLastPInvokeError(__lastError);
	return __retVal;
	[DllImport("libSystem.Native", EntryPoint = "SystemNative_GetGroupName", ExactSpelling = true)]
	[CompilerGenerated]
	static extern unsafe byte* __PInvoke__(uint uid);
}
```

Reply:
@AaronRobinsonMSFT @elinor-fung @jkoritzinsky - for my knowledge, if I have a P/Invoke where the returned string is shared and shouldn't be free'd, how would I write that? Just return the `byte*` pointer in my method and handle it myself? Or is there some built-in way to tell the marshalling code to not free the buffer?

Reply:
>  if I have a P/Invoke where the returned string is shared and shouldn't be free'd, how would I write that?

That would be a special case and I'm dubious if it is all that common. There is no support for that with the source generator marshallers nor in the built-in marshallers. Feel free to file an issue on that and we can consider it. Full-disclosure, if there isn't a lot of traction on it we'd likely close it at the end of the .NET 8 cycle.

> Just return the byte* pointer in my method and handle it myself?

That would be the easiest way to handle this case. Another way would be to copy the appropriate "StringMarshaller" type from source into the project and modify it as desired.